rust/lance-index/src/vector/hnsw/builder.rs

```diff
@@ -204,6 +206,46 @@ impl HNSW {
         Ok(hnsw)
     }
 
+    pub fn build_empty(distance_type: DistanceType, params: HnswBuildParams) -> Self {
+        let inner = HNSWBuilderInner::with_params(distance_type, params, 0);
+        Self {
+            inner: Arc::new(inner),
+        }
+    }
+
+    pub fn insert(
+        &self,
+        node: u32,
+        visited_generator: &mut VisitedGenerator,
+        storage: &impl VectorStore,
+    ) -> Result<()> {
+        if node as usize >= self.inner.nodes().read().unwrap().len() {
+            let mut write_nodes = self.inner.nodes.write().unwrap();
+            if node as usize >= write_nodes.len() {
+                let mut new_nodes: Vec<_> = write_nodes
+                    .iter()
+                    .map(|v| RwLock::new(v.read().unwrap().clone()))
+                    .collect();
+                let mut new_size = max(write_nodes.len(), 1);
+                while node as usize >= new_size {
+                    new_size *= 2;
```

Comment:
another small nit: Maybe we should make this configurable. something something python uses something very mild like 1.1125 https://github.com/python/cpython/blob/main/Objects/listobject.c#L109-L119

Reply:
Nodes take up very little memory compared to storage, so I'm gonna leave that for another PR. You might want to consider merging this extension with a similar thing for vector storage later (so that they happen at the same time).